mono/metadata/threads.c

```diff
@@ -5002,6 +5015,66 @@ mono_thread_is_foreign (MonoThread *thread)
 	return info->runtime_thread == FALSE;
 }
 
+#ifndef HOST_WIN32
+static void
+threads_native_thread_join_lock (gpointer key, gpointer value)
+{
+	pthread_t thread = (pthread_t)key;
+	if (thread != pthread_self ()) {
+		MONO_ENTER_GC_SAFE;
+		/* This shouldn't block */
+		mono_threads_join_lock ();
+		mono_native_thread_join (thread);
+		mono_threads_join_unlock ();
+		MONO_EXIT_GC_SAFE;
+	}
+}
+inline static void
+threads_native_thread_join_nolock (gpointer key, gpointer value)
+{
+	pthread_t thread = (pthread_t)key;
+	MONO_ENTER_GC_SAFE;
+	mono_native_thread_join (thread);
+	MONO_EXIT_GC_SAFE;
+}
+
+inline static void
+threads_add_joinable_thread_nolock (gpointer key)
+{
+	g_hash_table_insert (joinable_threads, key, key);
+}
+#else
+static void
+threads_native_thread_join_lock (gpointer key, gpointer value)
+{
```

Comment:
What do the key/value parameters mean here ?

Reply:
Key/Value will be different on Windows where key is the thread ID and value is the OS handle to the thread (preventing thread id to be recycled). On other platforms key will be the same as value (same as previous implementation where tid was both key and value), but same function signature is used on all platforms. So for Windows key == tid and value == thread handle, on other platforms key == tid and value == tid (the same as previous implementation). I can give the parameters different names if that makes it clearer.